experimental/waveshare_eink/linux_kernel_modules/fb_waveshare_eink.c

```diff
@@ -38,6 +38,8 @@
 #define WS_SET_DUMMY_LINE_PERIOD		0x3A
 #define WS_SET_GATE_TIME			0x3B
 
+#define MSB2LSB(b) (((b)&1?128:0)|((b)&2?64:0)|((b)&4?32:0)|((b)&8?16:0)|((b)&16?8:0)|((b)&32?4:0)|((b)&64?2:0)|((b)&128?1:0))
```

Comment:
I believe this macro reverses the bits in a byte.  So for example `MSB2LSB(0x3A) -> 0x5C // (0b00111010 -> 0b01011100)`

This is not the same concept as big/little endian.  Big endian means that bytes in integers wider than one byte are stored with the most significant byte in the lowest memory address.

So `u32 x = 0x12345678;` is stored in memory on a big endian architecture as:
```
low addr  | 0x12
          | 0x34
          | 0x56
high addr | 0x78
```

and on a little endian architecture it would be:
```
low addr  | 0x78
          | 0x56
          | 0x34
high addr | 0x12
```

Problems can arise when a CPU and a peripheral share memory and try to exchange multi byte integers. So if you knew that your peripheral that you are writing a driver for uses little endian, then you need to write code like this in your driver:
```
void write_resolution(u32 width, u32 height)
{
        width = cpu_to_le32(width);
        height = cpu_to_le32(height);
        my_device_write_bytes(dev, DEVICE_OFFSET_WIDTH, (u8 *)width, sizeof(width));
        my_device_write_bytes(dev, DEVICE_OFFSET_HEIGHT, (u8 *)height, sizeof(height));
}
```

From what I have seen in the datasheet, I don't *think* this driver should have any endian issues, but the code that you had originally did have some endian ifdefs and it's not clear to me why. Please do some analysis of the driver datasheet to decide if any endian conversion functions are necessary.

Reply:
Hi David

As I checked in TI doc with 16-bit pixel

The 16-bit pixel format presented an issue with userspace compatibility 
– All userspaceapplication assume that framebuffers are organized in a native endian format. 
– On our little endian ARM system, the mmaped shadow framebuffer is written in native little 
endian

–SPI buffer transfers in 8-bit data mode required by the ST7735 do a byte swap by nature of the byte-wise addressing of the PIO or DMA based memory access:
. Have to present the SPI adapter driver with a byte swapped shadow buffer 
. Driver has hack which byte swaps the buffer before doing a spi_write() on every deferred_io update. This allows unmodified use existing FB API applications

Reply:
For your question: The only other major issue that I can remember with the driver is that it doesn't yet make a distinction between the visible space (250x122) vs the addressable space (250x128)

So why we set 128 instead of 122. Because the x coordinates must be the multiple of 8.

Below is info from: https://www.waveshare.com/wiki/2.13inch_e-Paper_HAT

   Private function: Set the memory pointer (SetMemoryPointer)

   SetMemoryPointer is used to set the start point of the following image to be sent. Because 1 byte = 8 pixels of the image data to be sent, the x coordinates must be the multiple of 8, or else the last 3 bits will be ignored.

Reply:
Hi David

I checked datasheet and refer some sample framebuffer driver but there are no where mention about need to convert little endian and big endian in framebuffer. 
Beside I checked our framebuffer driver with another lcd(TFT 2.8 inch ILI9341). There are no issue with this lcd.

So you can ignore my request for checking endian and 3 commit you can help to merge are: 
c952d56: ws eink: use devm_gpio functions
fed5315: ws eink: remove endian-specific code 
16c897a: ws eink: Update Deferred IO delay 

For additional changes I will create new pull request.   

Thanks
Thong Nguyen
![20190130_163954](https://user-images.githubusercontent.com/17214533/51972977-42cb3080-24af-11e9-81be-80c65c0e2890.jpg)